### PR TITLE
Connor4312/terminal types contribution

### PR DESCRIPTION
**Pull request by connor4312**
For https://github.com/microsoft/vscode/issues/98054

Tried to mirror what I saw other parts of the workbench doing.

I tried to find a way to get back structured objects from the `SelectActionViewItem` so that we didn't need to implement string matching, but found this difficult to do; the cleanest way would be deriving an interface from the `ISelectOptionItem`, but the selection options aren't exposed back to the `SelectActionViewItem` at any point and the types in general always have the select only emitting a string.

Seems to work, and should be fine as long as people don't contribute a terminal action that starts with `[0-9]: `

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	public readonly _serviceBrand = undefined;
+
+	private _terminalTypes: ReadonlyArray<ITerminalTypeContribution> = [];
+
+	public get terminalTypes() {
+		return this._terminalTypes;
+	}
+
+	constructor() {
+		terminalsExtPoint.setHandler(contributions => {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts:32]: We need to protect usage of this for only when proposed API is enabled

File: src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// terminal extension point
+export const terminalsExtPoint = extensionsRegistry.ExtensionsRegistry.registerExtensionPoint<ITerminalContributions>(terminalContributionsDescriptor);
+
+export interface ITerminalContributionService {
+	readonly _serviceBrand: undefined;
+
+	readonly terminalTypes: ReadonlyArray<ITerminalTypeContribution>;
+}
+
+export const ITerminalContributionService = createDecorator<ITerminalContributionService>('terminalContributionsService');
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts:20]: I don't see any other `IXContributionService`s, what's the usual pattern for this sort of thing?

**Review by Tyriar — COMMENTED** [review]
Avoiding the string matching would definitely be nice, it was an existing problem though.

Just some minor comments, as it will be proposed I don't have problems with pushing this in once the comments are resolved. We should push to finalize it after it's been tested out though.

File: src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts (hunk 2 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts:20]: They seem to usually be gathered up in some contribution service, e.g. https://github.com/microsoft/vscode/blob/master/src/vs/workbench/contrib/codeActions/common/codeActionsContribution.ts#L54 or https://github.com/microsoft/vscode/blob/master/src/vs/workbench/contrib/tasks/common/taskDefinitionRegistry.ts#L88

Often the contribution services do additional things with the contribution information, but here there's no need for that yet.

File: src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts (hunk 1 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts:33]: Seems like the common pattern in vscode itself is to silently ignore contributions that require a proposed API to work, e.g. `themeExtensionPoints.ts` or `proxyResolver.ts`. Not sure if that's the best thing for extension authors, but followed that pattern here.

File: src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts (hunk 2 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/common/terminalExtensionPoints.ts:20]: Let's stick to this for now I guess, should be easy to change later as needed.

**Review by Tyriar — APPROVED** [review]

**Pull request merged**